### PR TITLE
Add max query depth

### PR DESCRIPTION
**Pull request by gnawf**
Ok I also removed the engine abstraction stuff and added `transforms` and `introspectionRunnerFactory` to `Nadel.Builder` instead of passing those directly to `NextgenEngine`.

Tested this new build with the Gateway and it works fine.

### CODE REVIEW

File: lib/build.gradle.kts [hunk 1]
```diff
@@ -7,7 +7,7 @@ plugins {
     id("com.bnorm.power.kotlin-power-assert")
 }
 
-val graphqlJavaVersion = "0.0.0-2023-06-07T00-38-50-5b19375"
+val graphqlJavaVersion = "0.0.0-2023-07-12T23-55-39-fa8cf1b"
```
Comment (gnawf) [inline lib/build.gradle.kts:10]: Brings in the max ENF depth change.

File: lib/src/main/java/graphql/nadel/Nadel.kt [hunk 2]
```diff
@@ -38,15 +41,12 @@ import java.util.concurrent.atomic.AtomicReference
 import java.util.function.Function
 
 class Nadel private constructor(
-    private val engine: NadelExecutionEngine,
+    private val engine: NextgenEngine,
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/Nadel.kt:44]: Removed the engine abstraction. Nadel itself creates the engine now. All constructor params that were tied to `NextgenEngine` are now in `Nadel.Builder`.

File: lib/src/main/java/graphql/nadel/Nadel.kt [hunk 3]
```diff
@@ -38,15 +41,12 @@ import java.util.concurrent.atomic.AtomicReference
...
     val services: List<Service>,
     val engineSchema: GraphQLSchema,
     val querySchema: GraphQLSchema,
-    internal val serviceExecutionFactory: ServiceExecutionFactory,
-    internal val instrumentation: NadelInstrumentation,
-    internal val serviceExecutionHooks: ServiceExecutionHooks,
-    internal val preparsedDocumentProvider: PreparsedDocumentProvider,
-    internal val executionIdProvider: ExecutionIdProvider,
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/Nadel.kt:49]: Removed some things that are no longer needed in this class. Most of the things are now directly passed down to `NextgenEngine`

File: lib/src/main/java/graphql/nadel/Nadel.kt [hunk 4]
```diff
@@ -226,15 +226,17 @@ class Nadel private constructor(
...
 
     class Builder {
-        private var serviceExecutionFactory: ServiceExecutionFactory? = null
         private var instrumentation: NadelInstrumentation = object : NadelInstrumentation {}
         private var serviceExecutionHooks: ServiceExecutionHooks = object : ServiceExecutionHooks {}
         private var preparsedDocumentProvider: PreparsedDocumentProvider = NoOpPreparsedDocumentProvider.INSTANCE
         private var executionIdProvider = ExecutionIdProvider.DEFAULT_EXECUTION_ID_PROVIDER
-        private var engineFactory: NadelExecutionEngineFactory = NadelExecutionEngineFactory(::NextgenEngine)
+        private var transforms = emptyList<NadelTransform<out Any>>()
+        private var introspectionRunnerFactory = NadelIntrospectionRunnerFactory(::NadelDefaultIntrospectionRunner)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/Nadel.kt:234]: These used to be constructor parameters to `NextgenEngine` but now are here

File: lib/src/main/java/graphql/nadel/Nadel.kt [hunk 5]
```diff
@@ -226,15 +226,17 @@ class Nadel private constructor(
...
 
         private var schemaBuilder = NadelSchemas.Builder()
 
+        private var maxQueryDepth = Integer.MAX_VALUE
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/Nadel.kt:238]: New option to set max ENF depth

File: lib/src/main/java/graphql/nadel/NextgenEngine.kt [hunk 6]
```diff
@@ -53,46 +56,49 @@ import kotlinx.coroutines.future.asCompletableFuture
 import kotlinx.coroutines.future.asDeferred
 import kotlinx.coroutines.future.await
 import kotlinx.coroutines.launch
-import java.util.Locale
 import java.util.concurrent.CompletableFuture
+import graphql.normalized.ExecutableNormalizedOperationFactory.Options.defaultOptions as executableNormalizedOperationFactoryOptions
 
-class NextgenEngine @JvmOverloads constructor(
-    nadel: Nadel,
+internal class NextgenEngine(
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/NextgenEngine.kt:62]: Most of this is just changes to the constructor now that the engine abstraction is gone.

The `Nadel.Builder` now just passes everything required directly in here, so we don't need to go `nadel.` to get everything out.

File: lib/src/main/java/graphql/nadel/NextgenEngine.kt [hunk 7]
```diff
@@ -53,46 +56,49 @@ import kotlinx.coroutines.future.asCompletableFuture
...
         introspectionRunnerFactory = introspectionRunnerFactory,
         dynamicServiceResolution = dynamicServiceResolution,
-        services,
+        services = this.services,
     )
-    private val executionIdProvider = nadel.executionIdProvider
 
-    override fun execute(
+    private val operationParseOptions = executableNormalizedOperationFactoryOptions()
+        .maxChildrenDepth(maxQueryDepth)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/NextgenEngine.kt:99]: Base options with the query depth set.

**Review by temaEmelyan — APPROVED** [review]
👍

**Review by bbakerman — APPROVED** [review]
Nice one

**Pull request merged**